*** a/slstatus.c	2017-03-16 07:38:28.050112122 -0500
--- b/slstatus.c	2017-03-16 19:42:24.337348057 -0500
***************
*** 103,121 ****
  battery_perc(const char *bat)
  {
  	int perc;
  	char path[PATH_MAX];
  	FILE *fp;
  
! 	snprintf(path, sizeof(path), "%s%s%s", "/sys/class/power_supply/", bat, "/capacity");
  	fp = fopen(path, "r");
  	if (fp == NULL) {
  		warn("Failed to open file %s", path);
  		return smprintf("%s", UNKNOWN_STR);
  	}
! 	fscanf(fp, "%i", &perc);
  	fclose(fp);
  
! 	return smprintf("%d%%", perc);
  }
  
  static char *
--- 103,146 ----
  battery_perc(const char *bat)
  {
  	int perc;
+     int charge_now;
+     int charge_full_design;
  	char path[PATH_MAX];
  	FILE *fp;
  
! 	snprintf(path, sizeof(path), "%s%s%s", "/sys/class/power_supply/", bat, "/charge_full_design");
  	fp = fopen(path, "r");
  	if (fp == NULL) {
  		warn("Failed to open file %s", path);
  		return smprintf("%s", UNKNOWN_STR);
  	}
!     fscanf(fp, "%i", &charge_full_design);
  	fclose(fp);
  
! 	snprintf(path, sizeof(path), "%s%s%s", "/sys/class/power_supply/", bat, "/charge_now");
! 	fp = fopen(path, "r");
! 	if (fp == NULL) {
! 		warn("Failed to open file %s", path);
! 		return smprintf("%s", UNKNOWN_STR);
! 	}
!     fscanf(fp, "%i", &charge_now);
! 	fclose(fp);
! 
!     perc = (int)(100.0*((double)charge_now/(double)charge_full_design));
! 
!     if (perc <= battery_urgent)
!     {
! 	    return smprintf("%d%%", perc);
!     }
!     else if (perc <= battery_low)
!     {
! 	    return smprintf("%d%%", perc);
!     }
!     else
!     {
!   	    return smprintf("%d%%", perc);
!     }
! 
  }
  
  static char *
***************
*** 135,145 ****
  	fclose(fp);
  
  	if (strcmp(state, "Charging") == 0) {
! 		return smprintf("+");
  	} else if (strcmp(state, "Discharging") == 0) {
! 		return smprintf("-");
  	} else if (strcmp(state, "Full") == 0) {
! 		return smprintf("=");
  	} else {
  		return smprintf("?");
  	}
--- 160,170 ----
  	fclose(fp);
  
  	if (strcmp(state, "Charging") == 0) {
!         return smprintf("\U0001F50C");
  	} else if (strcmp(state, "Discharging") == 0) {
! 		return smprintf("\U0001F50B");
  	} else if (strcmp(state, "Full") == 0) {
! 		return smprintf("F");
  	} else {
  		return smprintf("?");
  	}
***************
*** 186,192 ****
  		return smprintf("%s", UNKNOWN_STR);
  	}
  
! 	return smprintf("%s", str);
  }
  
  static char *
--- 211,217 ----
  		return smprintf("%s", UNKNOWN_STR);
  	}
  
! 	return smprintf("%s", str);
  }
  
  static char *
***************
*** 683,689 ****
  
  	close(afd);
  
! 	return smprintf("%d%%", v & 0xff);
  }
  
  static char *
--- 708,733 ----
  
  	close(afd);
  
!     if (v < 1) 
!     {
! 	    return smprintf("\U0001F507  %d%%", v & 0xff);
!     }
!     else if (v < 10)
!     {
! 	    return smprintf("\U0001F508  %d%%", v & 0xff);
!     }
!     else if (v < 30)
!     {
! 	    return smprintf("\U0001F508 %d%%", v & 0xff);
!     }
!     else if (v < 65)
!     {
! 	    return smprintf("\U0001F509 %d%%", v & 0xff);
!     }
!     else
!     {
! 	    return smprintf("\U0001F50A %d%%", v & 0xff);
!     }
  }
  
  static char *
***************
*** 725,731 ****
  	datastart = (datastart+(strlen(iface)+1));
  	sscanf(datastart + 1, " %*d   %d  %*d  %*d		  %*d	   %*d		%*d		 %*d	  %*d		 %*d", &perc);
  
! 	return smprintf("%d%%", perc);
  }
  
  static char *
--- 769,782 ----
  	datastart = (datastart+(strlen(iface)+1));
  	sscanf(datastart + 1, " %*d   %d  %*d  %*d		  %*d	   %*d		%*d		 %*d	  %*d		 %*d", &perc);
  
!     if (perc < wifi_low)
!     {
! 	    return smprintf("%d%%", perc);
!     }
!     else
!     {
! 	    return smprintf("%d%%", perc);
!     }
  }
  
  static char *
