*** a/slstatus.c	2017-04-07 08:20:23.755220179 -0500
--- b/slstatus.c	2017-04-07 08:10:46.159390544 -0500
***************
*** 63,68 ****
--- 63,69 ----
  static char *uptime(void);
  static char *username(void);
  static char *vol_perc(const char *card);
+ static char *vpn_status(const char *iface);
  static char *wifi_perc(const char *iface);
  static char *wifi_essid(const char *iface);
  static void sighandler(const int signo);
***************
*** 103,121 ****
  battery_perc(const char *bat)
  {
  	int perc;
  	char path[PATH_MAX];
  	FILE *fp;
  
! 	snprintf(path, sizeof(path), "%s%s%s", "/sys/class/power_supply/", bat, "/capacity");
  	fp = fopen(path, "r");
  	if (fp == NULL) {
  		warn("Failed to open file %s", path);
  		return smprintf("%s", UNKNOWN_STR);
  	}
! 	fscanf(fp, "%i", &perc);
  	fclose(fp);
  
! 	return smprintf("%d%%", perc);
  }
  
  static char *
--- 104,147 ----
  battery_perc(const char *bat)
  {
  	int perc;
+     int charge_now;
+     int charge_full_design;
  	char path[PATH_MAX];
  	FILE *fp;
  
! 	snprintf(path, sizeof(path), "%s%s%s", "/sys/class/power_supply/", bat, "/charge_full_design");
  	fp = fopen(path, "r");
  	if (fp == NULL) {
  		warn("Failed to open file %s", path);
  		return smprintf("%s", UNKNOWN_STR);
  	}
!     fscanf(fp, "%i", &charge_full_design);
  	fclose(fp);
  
! 	snprintf(path, sizeof(path), "%s%s%s", "/sys/class/power_supply/", bat, "/charge_now");
! 	fp = fopen(path, "r");
! 	if (fp == NULL) {
! 		warn("Failed to open file %s", path);
! 		return smprintf("%s", UNKNOWN_STR);
! 	}
!     fscanf(fp, "%i", &charge_now);
! 	fclose(fp);
! 
!     perc = (int)(100.0*((double)charge_now/(double)charge_full_design));
! 
!     if (perc <= battery_urgent)
!     {
! 	    return smprintf("%d%%", perc);
!     }
!     else if (perc <= battery_low)
!     {
! 	    return smprintf("%d%%", perc);
!     }
!     else
!     {
!   	    return smprintf("%d%%", perc);
!     }
! 
  }
  
  static char *
***************
*** 135,147 ****
  	fclose(fp);
  
  	if (strcmp(state, "Charging") == 0) {
! 		return smprintf("+");
  	} else if (strcmp(state, "Discharging") == 0) {
! 		return smprintf("-");
  	} else if (strcmp(state, "Full") == 0) {
! 		return smprintf("=");
! 	} else if (strcmp(state, "Unknown") == 0) {
! 		return smprintf("/");
  	} else {
  		return smprintf("?");
  	}
--- 161,171 ----
  	fclose(fp);
  
  	if (strcmp(state, "Charging") == 0) {
!         return smprintf("\U0001F50C");
  	} else if (strcmp(state, "Discharging") == 0) {
! 		return smprintf("\U0001F50B");
  	} else if (strcmp(state, "Full") == 0) {
! 		return smprintf("F");
  	} else {
  		return smprintf("?");
  	}
***************
*** 188,194 ****
  		return smprintf("%s", UNKNOWN_STR);
  	}
  
! 	return smprintf("%s", str);
  }
  
  static char *
--- 212,218 ----
  		return smprintf("%s", UNKNOWN_STR);
  	}
  
! 	return smprintf("%s", str);
  }
  
  static char *
***************
*** 447,455 ****
  		warn("Failed to get command output for %s", cmd);
  		return smprintf("%s", UNKNOWN_STR);
  	}
! 	fgets(buf, sizeof(buf) - 1, fp);
  	pclose(fp);
! 	buf[sizeof(buf) - 1] = '\0';
  
  	if ((nlptr = strstr(buf, "\n")) != NULL) {
  		nlptr[0] = '\0';
--- 471,479 ----
  		warn("Failed to get command output for %s", cmd);
  		return smprintf("%s", UNKNOWN_STR);
  	}
! 	fgets(buf, sizeof(buf), fp);
  	pclose(fp);
! 	buf[sizeof(buf)] = '\0';
  
  	if ((nlptr = strstr(buf, "\n")) != NULL) {
  		nlptr[0] = '\0';
***************
*** 473,479 ****
  		return smprintf("%s", UNKNOWN_STR);
  	}
  
! 	if ((bytes_read = fread(buf, sizeof(char), sizeof(buf) - 1, fp)) == 0) {
  		warn("swap_free: read error");
  		fclose(fp);
  		return smprintf("%s", UNKNOWN_STR);
--- 497,503 ----
  		return smprintf("%s", UNKNOWN_STR);
  	}
  
! 	if ((bytes_read = fread(buf, sizeof(char), sizeof(buf), fp)) == 0) {
  		warn("swap_free: read error");
  		fclose(fp);
  		return smprintf("%s", UNKNOWN_STR);
***************
*** 510,516 ****
  		return smprintf("%s", UNKNOWN_STR);
  	}
  
! 	if ((bytes_read = fread(buf, sizeof(char), sizeof(buf) - 1, fp)) == 0) {
  		warn("swap_perc: read error");
  		fclose(fp);
  		return smprintf("%s", UNKNOWN_STR);
--- 534,540 ----
  		return smprintf("%s", UNKNOWN_STR);
  	}
  
! 	if ((bytes_read = fread(buf, sizeof(char), sizeof(buf), fp)) == 0) {
  		warn("swap_perc: read error");
  		fclose(fp);
  		return smprintf("%s", UNKNOWN_STR);
***************
*** 551,557 ****
  		warn("Failed to open file /proc/meminfo");
  		return smprintf("%s", UNKNOWN_STR);
  	}
! 	if ((bytes_read = fread(buf, sizeof(char), sizeof(buf) - 1, fp)) == 0) {
  		warn("swap_total: read error");
  		fclose(fp);
  		return smprintf("%s", UNKNOWN_STR);
--- 575,581 ----
  		warn("Failed to open file /proc/meminfo");
  		return smprintf("%s", UNKNOWN_STR);
  	}
! 	if ((bytes_read = fread(buf, sizeof(char), sizeof(buf), fp)) == 0) {
  		warn("swap_total: read error");
  		fclose(fp);
  		return smprintf("%s", UNKNOWN_STR);
***************
*** 582,588 ****
  		warn("Failed to open file /proc/meminfo");
  		return smprintf("%s", UNKNOWN_STR);
  	}
! 	if ((bytes_read = fread(buf, sizeof(char), sizeof(buf) - 1, fp)) == 0) {
  		warn("swap_used: read error");
  		fclose(fp);
  		return smprintf("%s", UNKNOWN_STR);
--- 606,612 ----
  		warn("Failed to open file /proc/meminfo");
  		return smprintf("%s", UNKNOWN_STR);
  	}
! 	if ((bytes_read = fread(buf, sizeof(char), sizeof(buf), fp)) == 0) {
  		warn("swap_used: read error");
  		fclose(fp);
  		return smprintf("%s", UNKNOWN_STR);
***************
*** 685,691 ****
  
  	close(afd);
  
! 	return smprintf("%d%%", v & 0xff);
  }
  
  static char *
--- 709,752 ----
  
  	close(afd);
  
!     if ((v & 0xff) < 1) 
!     {
! 	    return smprintf("\U0001F507  %d%%", v & 0xff);
!     }
!     else if ((v & 0xff) < 10)
!     {
! 	    return smprintf("\U0001F508  %d%%", v & 0xff);
!     }
!     else if ((v & 0xff) < 30)
!     {
! 	    return smprintf("\U0001F508 %d%%", v & 0xff);
!     }
!     else if ((v & 0xff) < 65)
!     {
! 	    return smprintf("\U0001F509 %d%%", v & 0xff);
!     }
!     else
!     {
! 	    return smprintf("\U0001F50A %d%%", v & 0xff);
!     }
! }
! 
! static char *
! vpn_status(const char *iface)
! {
! 	char path[PATH_MAX];
! 	char status[5];
! 	FILE *fp;
! 	snprintf(path, sizeof(path), "%s%s%s", "/sys/class/net/", iface, "/operstate");
! 	fp = fopen(path, "r");
! 	if (fp == NULL) {
! 		return smprintf("\U0001F513");
! 	}
! 	fgets(status, 5, fp);
! 	fclose(fp);
! 	if(strcmp(status, "down\n") != 0) {
! 		return smprintf("\U0001F512");
! 	}
  }
  
  static char *
***************
*** 727,733 ****
  	datastart = (datastart+(strlen(iface)+1));
  	sscanf(datastart + 1, " %*d   %d  %*d  %*d		  %*d	   %*d		%*d		 %*d	  %*d		 %*d", &perc);
  
! 	return smprintf("%d%%", perc);
  }
  
  static char *
--- 788,801 ----
  	datastart = (datastart+(strlen(iface)+1));
  	sscanf(datastart + 1, " %*d   %d  %*d  %*d		  %*d	   %*d		%*d		 %*d	  %*d		 %*d", &perc);
  
!     if (perc < wifi_low)
!     {
! 	    return smprintf("%d%%", perc);
!     }
!     else
!     {
! 	    return smprintf("%d%%", perc);
!     }
  }
  
  static char *
